proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,397 @@
+     When the same package is transitively visible in two stages, there is no same-version constraint across the two stages that arises out of thin air.
+
+Effect and Interactions
+-----------------------
+
+Here is an example of many the features used together, rewriting the code from the motivation.
+Hypothetical ``ios-th`` package:
+::
+  {-# LANGUAGE TemplateHaskell #-}
+  {-# LANGUAGE NoTemplateStagePersistence #-}
+  module Ios.Macros where
+
+  #ifndef ios_TARGET_OS
+  # error Module shouldn't be built. Fix Cabal file!
+  #endif
+
+  import Language.Haskell.TH
+  $import 1 Ios.Types (Foo(..))
+
+  iosBoilerplateHelper :: Name -> Q Expr
+  iosBoilerplateHelper name = ... [| ... :: Foo |] ...
+end user code:
+::
+  {-# LANGUAGE TemplateHaskell #-}
+  {-# LANGUAGE NoTemplateStagePersistence #-}
+  module MyApp.Ios where
+
+  #ifndef ios_HOST_OS
+  # error Module shouldn't be built. Fix Cabal file!
+  #endif
+
+  import Ios.Types
+  $import -1 Ios.Macros
+
+  data SomeIosFfiType
+
+  $let -1 unneededBinding = iosBoilerplateHelper ''SomeIosFfiType
+
+  $(unneededBinding)
+
+A few misc implementation notes:
+
+Bindings interleave stages
+  Note that ``$let`` can appear outside the top-level, including in contexts where a variable of later stage is bound.
+  At first glance, binding a compile-time variable within a run-time variable might seem like a staging violation:
+  ::
+    f x = $huh
+       where foo = ...
+             $let -1 huh ... = ... [| x |] ... [| foo |] ...
+  But remember that later stage syntax can just be used in quotes; it is inert and cannot be evaluated.
+  ``huh`` is trivially lifted outside of ``f`` since captures the syntactic ``x`` which is static at compile-time.
+  Nothing passed into ``f`` at any call site is available to ``huh``.
+
+Forward references across splices
+   The intra-module staging restriction is gone, but that's separate from the prohibition on referencing bindings.
+   It just avoids the need to topologically sort splices based references from the quotations inside them, or break cycles à la ``*.hs-boot``.
+   Nevertheless, implementing that is not trivial so it is good to decouple relaxing the restriction from this already-large proposal.
+   Hopefully a future proposal will tackle this.
+
+Speeding up builds
+  Modules and libraries are relative in that their exported stage 0 may not be imported at stage 0.
+  There is no notion of a global "true" stage 0.
+  This is good in that we can share build artifacts without breaking abstractions.
+  For example, in the common native case, a library that needs another library in stage 0 and stage -1 can load the *same* build of the library in both of those stages.
+  By virtue of explicit stage of the import, the definitions do not unify even though the underlying build is the same.
+  This can be compared to repeated abstract interfaces in backpack being instantiated with the same concrete module.
```

Comment:
Possibly silly question: could we implement it in exactly that way? i.e. use the machinery we have for Backpack to handle the multiple "stage instantiations" of a module?

Reply:
Yes I hope! In general I want to leverage backpack in behind the scenes in a variety of ways. I think it's just better to type check against an interface and only later plug in the machine code.